File: conformance/tests/p256-primitives/p256-primitives.runar.ts
import { SmartContract, assert, p256Add, p256Mul, p256MulGen, p256OnCurve } from 'runar-lang';
import type { P256Point } from 'runar-lang';

class P256Primitives extends SmartContract {
  readonly expectedPoint: P256Point;

  constructor(expectedPoint: P256Point) {
    super(expectedPoint);
    this.expectedPoint = expectedPoint;
  }

  public verify(k: bigint, basePoint: P256Point) {
    const result: P256Point = p256Mul(basePoint, k);
    assert(p256OnCurve(result));
    assert(result == this.expectedPoint);
  }

  public verifyAdd(a: P256Point, b: P256Point) {
    const result: P256Point = p256Add(a, b);
    assert(p256OnCurve(result));
    assert(result == this.expectedPoint);
  }

  public verifyMulGen(k: bigint) {
    const result: P256Point = p256MulGen(k);
    assert(p256OnCurve(result));
    assert(result == this.expectedPoint);
  }
}

File: conformance/tests/p256-primitives/p256-primitives.runar.py
from runar import SmartContract, ByteString, Bigint, public, assert_, p256_mul, p256_add, p256_mul_gen, p256_on_curve

class P256Primitives(SmartContract):
    expected_point: ByteString

    def __init__(self, expected_point: ByteString):
        super().__init__(expected_point)
        self.expected_point = expected_point

    @public
    def verify(self, k: Bigint, base_point: ByteString):
        result = p256_mul(base_point, k)
        assert_(p256_on_curve(result))
        assert_(result == self.expected_point)

    @public
    def verify_add(self, a: ByteString, b: ByteString):
        result = p256_add(a, b)
        assert_(p256_on_curve(result))
        assert_(result == self.expected_point)

    @public
    def verify_mul_gen(self, k: Bigint):
        result = p256_mul_gen(k)
        assert_(p256_on_curve(result))
        assert_(result == self.expected_point)
